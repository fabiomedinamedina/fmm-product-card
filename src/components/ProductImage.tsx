import React, { CSSProperties, useContext } from 'react';

import { ProductContext } from '.';

import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}


export const ProductImage = ({ img, className = '', style }: Props) => {

  const { product } = useContext( ProductContext );
  let imageToShow: string;
  if( img ){
    imageToShow = img;
  }else if( product.img ){
    imageToShow = product.img;
  }else {
    imageToShow = noImage;
  }

  return (
    <img
      className={ `${styles.productImg} ${ className }` }
      src={ imageToShow }
      alt="Product Image"
      style={ style }
    />
  );
};

