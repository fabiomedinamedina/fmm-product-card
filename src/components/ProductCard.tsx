import React, { CSSProperties, createContext } from 'react';

import { useProducts } from '../hooks';
import { InitialValues, Product, ProductCardHOCProps, ProductCardHandlers, ProductContextProps, onChangeArgs } from '../interfaces/interfaces';


import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;


export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: ( args: ProductCardHandlers ) => JSX.Element ;
  className?: string;
  style?: CSSProperties;
  onChange?: ( args: onChangeArgs ) => void;
  value?:number;
  initialValues?: InitialValues 
}

const ProductCardHOC = ({ children, product, className, style, onChange, value, initialValues }:Props) => {

  const { 
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset
  } = useProducts({ onChange, product, value, initialValues });
  
  return (
    <Provider value={{
      counter,
      increaseBy,
      maxCount,
      product
    }}>
      <div
        className={ `${ styles.productCard  } ${ className }` }
        style={style}
      >
        {
          children({
            count: counter,
            isMaxCountReached,
            maxCount,
            product,
            increaseBy,
            reset
          })
        }
      </div>
    </Provider>
  );
};

export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons
});