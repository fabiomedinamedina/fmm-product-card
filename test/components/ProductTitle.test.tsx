import React from 'react';
// import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { ProductTitle, ProductCard } from '../../src/components';
import { product } from '../data/products';

describe('Testing ProductTitle', () => {
  
  test('Debería mostrar el titulo correctamente', () => {
    const { container } = render( <ProductTitle title='Producto Prueba' />  );
   
    expect( container ).toMatchSnapshot();
    expect( screen.getByText( 'Producto Prueba' ) ).toBeTruthy();

  });

  test('Debería mostrar la clase correctamente', () => {

    render( <ProductTitle className='class-test' title='Producto de prueba' /> );
    const title = screen.getByText( 'Producto de prueba', { selector: '.class-test' } );
    expect( title ).toBeTruthy();
    
  });

  test('Debería mostrar el componente con el nombre del producto', () => {
    
    render(
      <ProductCard product={ product }>
        {
          () => (
            <ProductTitle />
          )
        }
      </ProductCard>
    );

    expect( screen.getByText( product.title ).innerHTML ).toBeTruthy();

  });

})