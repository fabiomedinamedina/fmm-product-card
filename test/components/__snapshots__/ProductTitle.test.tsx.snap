// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ProductTitle Deber√≠a mostrar el titulo correctamente 1`] = `
<div>
  <span
    class="productDescription "
  >
    Producto Prueba
  </span>
</div>
`;
